File: agent/verger.py
"""Verger: the front-desk agent for a volunteer-run org.

One round = read the inbox, answer what should be answered, and attempt the
sends. Every send passes the gate hook: allowlist check first, then a Strands
interrupt pauses the whole loop until the trustee decides. Interrupt state
persists in the session, so the trustee can decide hours later from a fresh
process (verger decide).
"""

import json
import os
import uuid

from openai import OpenAI
from strands import Agent, tool
from strands.hooks import BeforeToolCallEvent, HookProvider, HookRegistry
from strands.models.openai import OpenAIModel
from strands.session.file_session_manager import FileSessionManager

from agent import mail
from agent.ledger import Ledger
from agent.policy import ALLOWLIST, ORG_NAME, SYSTEM_PROMPT, gate

DATA_DIR = os.environ.get("VERGER_DATA", os.path.join(os.path.dirname(__file__), "..", ".data"))

_seen: set[str] = set()  # processed Message-IDs for O(1) membership across the round
ledger = Ledger(os.path.join(DATA_DIR, "receipts.jsonl"))


@tool
def read_inbox() -> list[dict]:
    """List the messages sitting in the org inbox."""
    msgs = mail.list_inbox()
    ledger.append(
        "inbox_read",
        {"count": len(msgs), "senders": sorted({m["from"] for m in msgs})},
    )
    return msgs


@tool
def read_message(message_id: str) -> str:
    """Read the full plain-text body of one inbox message by its id."""
    return mail.read_message(message_id)


@tool
def send_mail(to: str, subject: str, body: str) -> str:
    """Send one reply from the org desk. Every send passes the trustee gate."""
    mid = mail.send_mail(to, subject, body)
    ledger.append("mail_sent", {"to": to, "subject": subject, "message_id": mid})
    return f"sent to {to} (message-id {mid})"


class TrusteeGate(HookProvider):
    """The product's core promise, enforced at the tool-call layer.

    Hard policy first (allowlist); then a Strands interrupt pauses the loop so
    a human decides. Nothing leaves the org without that decision, and the
    decision can arrive hours later from a fresh process.
    """

    def register_hooks(self, registry: HookRegistry, **kwargs) -> None:
        registry.add_callback(BeforeToolCallEvent, self.check)

    def check(self, event: BeforeToolCallEvent) -> None:
        if event.tool_use["name"] != "send_mail":
            return
        to = event.tool_use["input"].get("to", "")
        subject = event.tool_use["input"].get("subject", "")
        ok, why = gate(to)
        if not ok:
            event.cancel_tool = f"BLOCKED BY POLICY: {why}"
            ledger.append("send_blocked", {"to": to, "subject": subject, "why": why})
            return
        decision = event.interrupt(
            "trustee-send-approval",
            reason={"to": to, "subject": subject, "body": event.tool_use["input"].get("body", "")[:400]},
        )
        if decision.strip().lower() not in ("approve", "y", "yes"):
            event.cancel_tool = f"Trustee did not approve the send to {to}."
            ledger.append("send_denied", {"to": to, "subject": subject})
        else:
            ledger.append("send_approved", {"to": to, "subject": subject})


def build_agent(session_id: str) -> Agent:
    key = os.environ["GROQ_API_KEY"]
    model = OpenAIModel(
        client_args={
            "api_key": key,
            "base_url": "https://api.groq.com/openai/v1",
        },
        model_id=os.environ.get("VERGER_MODEL", "openai/gpt-oss-120b"),
        params={"max_tokens": 1600, "temperature": 0.4},
    )
    return Agent(
        name="verger",
        model=model,
        system_prompt=SYSTEM_PROMPT,
        tools=[read_inbox, read_message, send_mail],
        hooks=[TrusteeGate()],
        callback_handler=None,
        session_manager=FileSessionManager(session_id=session_id, storage_dir=os.path.join(DATA_DIR, "session")),
    )


def _pending_path() -> str:
    return os.path.join(DATA_DIR, "pending-interrupts.json")


def _load_pending() -> dict:
    path = _pending_path()
    if os.path.exists(path):
        with open(path) as f:
            return json.load(f)
    return {}


def _save_pending(items: dict) -> None:
    with open(_pending_path(), "w") as f:
        json.dump(items, f, indent=2, sort_keys=True)


def _session_path() -> str:
    return os.path.join(DATA_DIR, "current-session")


def _current_session() -> str | None:
    if os.path.exists(_session_path()):
        with open(_session_path()) as f:
            return f.read().strip() or None
    return None


def _start_session() -> str:
    sid = f"round-{uuid.uuid4().hex[:10]}"
    with open(_session_path(), "w") as f:
        f.write(sid)
    return sid


def run_round(agent: Agent) -> dict:
    """Run one round until it ends or pauses on trustee decisions."""
    result = agent(f"It's Monday morning. Do your round for {ORG_NAME}.")
    pending = _load_pending()
    rounds = 0
    while result.stop_reason == "interrupt":
        rounds += 1
        for intr in result.interrupts:
            pending[intr.id] = {
                "id": intr.id,
                "name": intr.name,
                "reason": intr.reason,
            }
            ledger.append("send_awaiting", {"interrupt_id": intr.id, "reason": intr.reason})
        _save_pending(pending)
        break  # unattended run ends here; the trustee decides on their time
    if not pending and result.stop_reason != "interrupt":
        # the round completed: archive the session so the next round starts fresh
        os.remove(_session_path())
    return {
        "stop_reason": str(result.stop_reason),
        "pending_total": len(pending),
        "paused_this_run": rounds > 0,
    }


def start_round() -> dict:
    """Public entry for the CLI: refuse a new round while one is paused, else start fresh."""
    pending = _load_pending()
    if pending:
        return {
            "stop_reason": "refused",
            "pending_total": len(pending),
            "note": "a round is paused awaiting trustee decisions; decide first",
        }
    sid = _start_session()
    return run_round(build_agent(sid))


def resume_with(decision: str, interrupt_id: str | None = None) -> dict:
    """Trustee decided. Resume the session, feeding every interrupt a response."""
    sid = _current_session()
    if not sid:
        return {"stop_reason": "no-active-round", "still_pending": len(_load_pending())}
    agent = build_agent(sid)
    pending = _load_pending()
    targets = [interrupt_id] if interrupt_id else list(pending.keys())
    responses = [
        {"interruptResponse": {"interruptId": iid, "response": decision}}
        for iid in targets
    ]
    for iid in targets:
        pending.pop(iid, None)
    result = agent(responses)
    _save_pending(pending)
    while result.stop_reason == "interrupt":
        pending = _load_pending()
        for intr in result.interrupts:
            pending[intr.id] = {"id": intr.id, "name": intr.name, "reason": intr.reason}
            ledger.append("send_awaiting", {"interrupt_id": intr.id, "reason": intr.reason})
        _save_pending(pending)
        break
    if not pending and result.stop_reason != "interrupt":
        os.remove(_session_path())
    return {"stop_reason": str(result.stop_reason), "still_pending": len(pending)}

File: web/lib/verger.ts
import { Agent, tool, BeforeToolCallEvent } from "@strands-agents/sdk";
import { OpenAIModel } from "@strands-agents/sdk/models/openai";
import { z } from "zod";
import {
  addSent,
  getMailbox,
  getPending,
  getRoundMessages,
  setMailbox,
  setPending,
  setRoundMessages,
  type MailMessage,
} from "./blobs";
import { appendReceipt, markHandled } from "./ledger";
import type { PendingItem } from "./types";

/*
  The Verger agent, serverless edition: the Strands TypeScript SDK running
  inside one Netlify function invocation per inbox message (~10s budget).

  The round is chunked: the caller (the desk, or the scheduled function) keeps
  invoking processNextMessage until it reports done or held. Conversation
  continuity rides in Blobs, so chunk N remembers what chunk N-1 answered.

  The gate is the product: a BeforeToolCallEvent hook checks the allowlist and
  raises a Strands interrupt on every send. The interrupt surfaces as
  stopReason 'interrupt'; the trustee decides; the send executes
  deterministically in /api/decide. A quiet round is the success state.
*/

export const ORG_NAME = "Riverbank Community Fridge";

const ALLOWLIST: Record<string, string> = {
  "volunteer.maya@example.org": "volunteer",
  "elder.housing@example.org": "partner org",
  "donor.john@example.org": "donor",
  "events@example.org": "hall enquiries",
};

const SYSTEM_PROMPT = `You are Verger, the front-desk agent for ${ORG_NAME} (desk@foodbank.local), a volunteer-run community food project. You keep the org's inbox alive so trustees don't have to.

You handle ONE inbox message per turn. For the message you were given:
1. read_message to see it in full if you need more than the summary.
2. Decide: routine question, judgment call, or spam.
3. Routine: send_mail with a warm, short, specific reply (under 120 words).
4. Judgment call (dates, money, access, safety): still send_mail, and say in one leading sentence that the trustee will confirm.
5. FACTS POLICY (absolute): never state a time, date, quantity, price, name, or CAPABILITY that was not in the message or given to you. This includes promises like "we do deliver" or "shifts start at 9" unless the message said so. For any such detail, write that the trustee will confirm it. An invented detail is the worst thing you can do.
6. Spam: call no_reply and move on.

You reply to people, never lecture them.`;

export type ChunkResult =
  | { kind: "done"; note: string }
  | { kind: "held"; pendingId: string; to: string; subject: string }
  | { kind: "handled"; how: string; from: string; subject: string };

function buildModel() {
  const apiKey = process.env.GROQ_API_KEY;
  if (!apiKey) throw new Error("GROQ_API_KEY missing");
  return new OpenAIModel({
    api: "chat",
    modelId: process.env.VERGER_MODEL ?? "openai/gpt-oss-20b",
    apiKey,
    clientConfig: { baseURL: "https://api.groq.com/openai/v1" },
  });
}

export async function processNextMessage(): Promise<ChunkResult> {
  const mailbox = await getMailbox();
  const next = mailbox.find((m: MailMessage) => !m.handled);
  if (!next) {
    await setRoundMessages([]);
    return { kind: "done", note: "inbox clear" };
  }

  let heldPendingId: string | null = null;
  let heldInfo = { to: "", subject: "" };

  const readMessage = tool({
    name: "read_message",
    description: "Read the full text of the inbox message you are handling.",
    inputSchema: z.object({}),
    callback: async () => {
      const fresh = (await getMailbox()).find((m) => m.id === next.id);
      return fresh
        ? { from: fresh.from, subject: fresh.subject, body: fresh.body }
        : { error: "message vanished" };
    },
  });

  const sendMail = tool({
    name: "send_mail",
    description:
      "Send one reply from the org desk. Every send passes the trustee gate.",
    inputSchema: z.object({
      to: z.string().describe("recipient email address"),
      subject: z.string().describe("reply subject line"),
      body: z.string().describe("plain-text reply body"),
    }),
    callback: async (input: { to: string; subject: string; body: string }) => {
      // The gate hook raises the interrupt; this body only runs once approved.
      const messageId = `verger-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
      await addSent({
        to: input.to,
        subject: input.subject,
        body: input.body,
        sentAt: Date.now(),
        messageId,
        via: "netlify-demo-transport",
      });
      await markHandled(next.id, `replied: ${input.subject}`);
      await appendReceipt("mail_sent", {
        to: input.to,
        subject: input.subject,
        message_id: messageId,
      });
      return `sent to ${input.to}`;
    },
  });

  const noReply = tool({
    name: "no_reply",
    description:
      "Record that this message deliberately gets no reply (spam, abuse, nothing to act on).",
    inputSchema: z.object({ reason: z.string().describe("one short line") }),
    callback: async (input: { reason: string }) => {
      await markHandled(next.id, `no reply: ${input.reason}`);
      await appendReceipt("no_reply", {
        from: next.from,
        subject: next.subject,
        reason: input.reason,
      });
      return "recorded";
    },
  });

  const agent = new Agent({
    model: buildModel(),
    systemPrompt: SYSTEM_PROMPT,
    tools: [readMessage, sendMail, noReply],
    printer: false,
    messages: (await getRoundMessages()) as never,
  });

  // The gate: allowlist is the hard floor, then a Strands interrupt pauses the
  // send until the trustee decides. Registered post-construction via addHook.
  agent.addHook(BeforeToolCallEvent, (event) => {
    if (event.toolUse.name !== "send_mail") return;
    const input = event.toolUse.input as {
      to?: string;
      subject?: string;
      body?: string;
    };
    const to = String(input?.to ?? "").trim();
    if (!(to in ALLOWLIST)) {
      event.cancel = `BLOCKED BY POLICY: recipient ${to} is not on the trustee allowlist`;
      void appendReceipt("send_blocked", {
        to,
        subject: String(input?.subject ?? ""),
      });
      return;
    }
    event.interrupt({
      name: "trustee-send-approval",
      reason: {
        to: input.to ?? "",
        subject: input.subject ?? "",
        body: (input.body ?? "").slice(0, 400),
      },
    });
  });

  const result = await agent.invoke(
    `Handle this inbox message. From: ${next.from} | Subject: ${next.subject}`,
  );

  // carry the conversation forward for the next chunk of this round
  await setRoundMessages(agent.messages);

  if (result.stopReason === "interrupt") {
    const pending = await getPending();
    for (const intr of result.interrupts ?? []) {
      const reason = (intr.reason ?? {}) as {
        to?: string;
        subject?: string;
        body?: string;
      };
      heldPendingId = intr.id;
      heldInfo = { to: reason.to ?? "", subject: reason.subject ?? "" };
      pending[intr.id] = {
        id: intr.id,
        name: intr.name,
        reason,
      };
      await appendReceipt("send_awaiting", {
        interrupt_id: intr.id,
        to: reason.to,
        subject: reason.subject,
      });
    }
    await setPending(pending);
    await markHandled(next.id, `held at the gate: ${heldInfo.subject}`);
    return {
      kind: "held",
      pendingId: heldPendingId ?? "",
      to: heldInfo.to,
      subject: heldInfo.subject,
    };
  }

  // if the model finished without acting, do not silently strand the message
  const fresh = (await getMailbox()).find((m) => m.id === next.id);
  if (fresh && !fresh.handled) {
    await markHandled(next.id, "agent closed without action");
    await appendReceipt("no_reply", {
      from: next.from,
      subject: next.subject,
      reason: "agent closed without acting",
    });
  }

  if (result.stopReason === "cancelled_tool") {
    return {
      kind: "handled",
      how: "blocked by policy",
      from: next.from,
      subject: next.subject,
    };
  }
  return {
    kind: "handled",
    how: "agent handled it",
    from: next.from,
    subject: next.subject,
  };
}

// Deterministic trustee decision: approve sends the EXACT approved draft, deny
// records the refusal; either way the message is handled and the round continues.
export async function executeDecision(
  decision: string,
  interruptId: string | null,
): Promise<{ stop_reason: string; still_pending: number }> {
  const pending = await getPending();
  const id = interruptId ?? Object.keys(pending)[0];
  if (!id || !pending[id]) return { stop_reason: "nothing-pending", still_pending: Object.keys(pending).length };
  const item = pending[id];
  const to = item.reason.to ?? "";
  const subject = item.reason.subject ?? "";

  if (!(to in ALLOWLIST)) {
    await appendReceipt("send_blocked", { to, subject, why: "not on allowlist", interrupt_id: id });
    delete pending[id];
    await setPending(pending);
    await markHandledByTo(to, "blocked by allowlist");
    return { stop_reason: "blocked", still_pending: Object.keys(pending).length };
  }

  if (decision.trim().toLowerCase() in { approve: 1, y: 1, yes: 1 }) {
    await appendReceipt("send_approved", { to, subject, interrupt_id: id });
    const messageId = `verger-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
    await addSent({
      to,
      subject,
      body: item.reason.body ?? "",
      sentAt: Date.now(),
      messageId,
      via: "netlify-demo-transport",
    });
    await appendReceipt("mail_sent", { to, subject, message_id: messageId });
    await markHandledByTo(to, `replied: ${subject}`);
    await deletePending(pending, id);
    return { stop_reason: "sent", still_pending: Object.keys(pending).length };
  }

  await appendReceipt("send_denied", { to, subject, interrupt_id: id });
  await markHandledByTo(to, `denied: ${subject}`);
  await deletePending(pending, id);
  return { stop_reason: "denied", still_pending: Object.keys(pending).length };
}

async function markHandledByTo(to: string, note: string): Promise<void> {
  const mailbox = await getMailbox();
  const m = mailbox.find(
    (x) => !x.handled && x.from === to, // the pending item's `to` is the original sender
  );
  if (m) {
    m.handled = true;
    m.handleNote = note;
    await setMailbox(mailbox);
  }
}

async function deletePending(
  pending: Record<string, PendingItem>,
  id: string,
): Promise<void> {
  delete pending[id];
  await setPending(pending);
}
